service/src/main/kotlin/fi/espoo/evaka/invoicing/service/OutdatedIncomeNotifications.kt

```diff
@@ -145,23 +150,25 @@ class OutdatedIncomeNotifications(
                 traceId = msg.guardianId.toString(),
             )
             ?.also { emailClient.send(it) }
+            .also { db.transaction { it.createIncomeNotification(msg.guardianId, msg.type) } }
+    }
 
+    fun createExpiredIncome(
+        db: Database.Connection,
+        clock: EvakaClock,
+        msg: AsyncJob.CreateExpiredIncome,
+    ) {
         db.transaction {
-            it.createIncomeNotification(msg.guardianId, msg.type)
-
-            val firstDayAfterExpiration = clock.today().plusDays(1)
-            if (
-                msg.type == IncomeNotificationType.EXPIRED_EMAIL &&
-                    !it.personHasActiveIncomeOnDate(msg.guardianId, firstDayAfterExpiration)
-            ) {
+            val dayAfterExpiration = msg.incomeExpirationDate.plusDays(1)
+            if (!it.personHasActiveIncomeOnDate(msg.guardianId, dayAfterExpiration)) {
                 it.insertIncome(
                     clock = clock,
                     mapper = mapper,
                     income =
                         IncomeRequest(
                             personId = msg.guardianId,
                             effect = IncomeEffect.INCOMPLETE,
-                            validFrom = firstDayAfterExpiration,
+                            validFrom = dayAfterExpiration,
```

Comment:
you could consider (pseudo code)
validTo = tx.getIncomes(personId)
   .filter validFrom > dayAfterExpiration
   .minByOrNull { it.validFrom }
   ?.let { it minus days 1 }

Reply:
and for perf optimization the personHasActiveIncomeOnDate call could be replaced by searching the same incomes list